sdk/core/azure-core/inc/http/http.hpp

```diff
@@ -63,6 +66,191 @@ enum class HttpMethod
+   * and remove it from url
+   */
+  std::string parseUrl(std::string url)
+  {
+    auto position = url.find('?');
+
+    if (position == std::string::npos)
+    {
+      return url; // no query parameters. Nothing else to do
+    }
+
+    // Get query parameters string and update
+    auto queryParameters = url.substr(position + 1);
+    url = url.substr(0, position);
+
+    // Split all query parameters (should be separated by &)
+    auto queryParametersVector = Request::split(queryParameters, '&');
+
+    // insert each query parameter to internal field
+    std::for_each(queryParametersVector.begin(), queryParametersVector.end(), [&](auto query) {
+      auto parameter = split(query, '=');
+      // Will throw if parameter in query is not valid (i.e. arg:1)
+      _queryParameters.insert(std::pair<std::string, std::string>(parameter[0], parameter[1]));
+    });
+
+    return url;
+  }
+
+  Request(
+      HttpMethod httpMethod,
+      std::string const& url,
+      BodyStream* bodyStream,
+      BodyBuffer* bodyBuffer)
+      : _method(std::move(httpMethod)), _url(parseUrl(std::move(url))), _bodyStream(bodyStream),
+        _bodyBuffer(bodyBuffer), _retryModeEnabled(false)
+  {
+    // TODO: parse url
+  }
+
+public:
+  Request(Request const&) = default;
+  Request& operator=(Request const&) = default;
+
+  Request(HttpMethod httpMethod, std::string const& url)
+      : Request(httpMethod, url, BodyStream::null, BodyBuffer::null)
+  {
+  }
+
+  Request(HttpMethod httpMethod, std::string const& url, BodyBuffer* bodyBuffer)
+      : Request(httpMethod, url, BodyStream::null, bodyBuffer)
+  {
+  }
+
+  Request(HttpMethod httpMethod, std::string const& url, BodyStream* bodyStream)
+      : Request(httpMethod, url, bodyStream, BodyBuffer::null)
+  {
+  }
+
+  // Methods used to build HTTP request
+  void addPath(std::string const& path);
+  void addQueryParameter(std::string const& name, std::string const& value);
+  void addHeader(std::string const& name, std::string const& value);
+  void startRetry(); // only called by retry policy
+
+  // Methods used by transport layer (and logger) to send request
+  HttpMethod getMethod();
+  std::string getEncodedUrl(); // should return encoded url
+  std::map<std::string, std::string> getHeaders();
+  BodyStream* getBodyStream();
+  BodyBuffer* getBodyBuffer();
+};
+
+/*
+ * Response exceptions
+ */
+struct CouldNotResolveHostException : public std::exception
+{
+  const char* what() const throw() { return "couldnt resolve host"; }
+};
+struct ErrorWhileWrittingResponse : public std::exception
+{
+  const char* what() const throw() { return "couldnt write response"; }
+};
+
+class Response
+{
+
+private:
+  uint16_t _statusCode;
+  std::string _reasonPhrase;
+  std::map<std::string, std::string> _headers;
+
+  // Response can contain no body, or either of next bodies (_bodyBuffer plus size or bodyStream)
+  http::BodyBuffer* _bodyBuffer;
+  http::BodyStream* _bodyStream;
```

Comment:
How do you write the response body to this stream? There isn't a `Write()` interface for `BodyStream` yet as far as I know.

Reply:
yeah, that not implemented yet.
this code returns a generic OK response for now.
I will add curl callbacks functions next.  
Just want reviewers to have it easier on not too big PRs